osc-server/src/main/java/org/osc/core/broker/util/crypto/X509TrustManagerFactory.java

```diff
@@ -220,31 +240,101 @@ private void reloadTrustManager() throws Exception {
         }
         reloadTrustManager();
         notifyTruststoreChanged();
+
     }
 
     @Override
-    public void addEntry(X509Certificate certificate, String newAlias) throws Exception {
-        if (fingerprintNotExist(getSha1Fingerprint(certificate))) {
-            this.keyStore.setCertificateEntry(newAlias, certificate);
-            try (FileOutputStream outputStream = new FileOutputStream(TRUSTSTORE_FILE)) {
-                this.keyStore.store(outputStream, getTruststorePassword());
-            }
-            reloadTrustManager();
-            notifyTruststoreChanged();
-        } else {
-            throw new Exception("Given certificate fingerprint already exists in trust store");
+    public void replaceInternalCertificate(File zipFile, boolean doReboot) throws Exception {
+        // Disabling reboot for testing.
+
+        if (zipFile == null) {
+            throw new Exception(BAD_KEY_PAIR_ZIP_FILE);
+        }
+
+        Map<String, File> files = unzipFilePair(zipFile);
+        File pKeyFile = files.get(KEY);
+        File chainFile = files.get(CERT_CHAIN);
+
+        FileUtils.deleteQuietly(zipFile);
+
+        replaceInternalCertificate(pKeyFile, chainFile, doReboot);
+    }
+
+    public void replaceInternalCertificate(File pKeyFile, File chainFile, boolean doReboot) throws Exception {
+        if (pKeyFile == null || chainFile == null) {
+            throw new Exception(BAD_KEY_PAIR_ZIP_FILE);
+        }
+
+        LOG.info(format("Replacing internal private/public keys from files %s and %s.",
+                 pKeyFile.getName(), chainFile.getName()));
+
+        Certificate[] internalCertificateChain;
+        Key internalKey;
+        // Loads the .pkipath certificate chain file
+        CertificateFactory cf = CertificateFactory.getInstance("X.509");
+        try (FileInputStream inputStream = new FileInputStream(chainFile)) {
+            CertPath certPath = cf.generateCertPath(inputStream);
+            List<? extends Certificate> certList = certPath.getCertificates();
+            internalCertificateChain = certList.toArray(new Certificate[]{});
+        }
+
+        BufferedReader br = new BufferedReader(new FileReader(pKeyFile));
+        PEMParser pp = new PEMParser(br);
+        PrivateKeyInfo pkInfo = (PrivateKeyInfo) pp.readObject();
+        internalKey = new JcaPEMKeyConverter().getPrivateKey(pkInfo);
+        pp.close();
+
+        try {
+            this.keyStore.setKeyEntry(INTERNAL_ALIAS, internalKey, getTruststorePassword(), internalCertificateChain);
+        } catch (KeyStoreException e) {
+            throw new Exception("Failed to add persist the new internal certificate!", e);
+        }
+
+        try(FileOutputStream outputStream = new FileOutputStream(TRUSTSTORE_FILE)) {
+            this.keyStore.store(outputStream, getTruststorePassword());
+        }
+
+        reloadTrustManager();
+        FileUtils.deleteQuietly(pKeyFile);
```

Comment:
we should probably keep these calls in a finally block...make sure all cleans of files are handled this way.